File: private_encapsulation.py
# Encapsulation bundles data (attributes) and methods (functions) together within a class, restricting external access to certain attributes.

# Private Access means members of a class are not accessible from outside the class. In Python, private members are denoted by double underscore (__) before the member name. but we can set and get the private variable using setter and getter methods.

class MyClass:
    def __init__(self):
        self.__privateVar = 0

    def setVar(self, privateVar):
        self.__privateVar = privateVar

    def getVar(self):
        return self.__privateVar


obj = MyClass()
obj.setVar(10)
print(obj.getVar())

# print(obj.__privateVar) # AttributeError: 'MyClass' object has no attribute '__privateVar'
# This is because __privateVar is a private variable and cannot be accessed from outside the class.

File: private_encapsulation.cpp
// Encapsulation bundles data (attributes) and methods (functions) together within a class, restricting external access to certain attributes.

// Private Access means members of a class are not accessible from outside the class. Only the class and friends can access private members. but we can set and get the value of private members using public member functions.

#include <bits/stdc++.h>
using namespace std;

class MyClass
{
private:
    int privateVar;

public:
    void setPrivateVar(int x)
    {
        privateVar = x;
    }
    int getPrivateVar()
    {
        return privateVar;
    }
};

int main()
{
    MyClass obj;
    obj.setPrivateVar(42);
    cout << obj.getPrivateVar() << endl; // 42
    // obj.privateVar = 42; // Error: 'int MyClass::privateVar' is private within this context
    return 0;
}

// In the above code, we have a class MyClass with a private data member privateVar. The privateVar member is not accessible from outside the class. We have defined public member functions setPrivateVar() and getPrivateVar() to set and get the value of the privateVar member. These member functions can be used to access and modify the privateVar member from outside the class. This is an example of encapsulation, where the data (privateVar) and methods (setPrivateVar() and getPrivateVar()) are bundled together within the class MyClass. The privateVar member is encapsulated within the class and can only be accessed through the public member functions. This helps in controlling access to the data and enforcing data integrity. Encapsulation is a fundamental concept in object-oriented programming and helps in building robust and maintainable code.
